text/0039-source-removal-buildpack.md

```diff
@@ -0,0 +1,61 @@
+# Source Removal Buildpack
+
+## Summary
+
+Adopt [ForestEckhardt's Source Removal buildpack](https://github.com/ForestEckhardt/source-removal)
+into the Paketo Project to allow for uniform source removal logic across all
+buildpacks that require such functionality.
```

Comment:
I like the concept & agree that implementing the logic in a consistent way across all buildpacks that require this in Paketo would be helpful.

What's challenging is that the way this is presently implemented for Java-related buildpack (and Rust), is that the delete has to happen at a specific time because there is work that happens after the deletion.

At a high level:

- Do a build, the artifact is stored in a layer so the build is cached
- We delete source code
- Artifact is taken from layer and moved to `/workspace`

I don't think we could change this for the Java buildpacks because buildpacks that run after our build tools, require the build artifact to exist in `/workspace`. This is because a user can push either code or a build artifact. If a user pushes a build artifact, it'll be in `/workspace`. Thus to build from source, we need our build tools buildpacks to compile and place the build artifact in the same location.

I could possibly see this buildpack working for the Rust buildpacks because they don't have the same need as Java.

How would you feel about exposing this buildpack as both a buildpack and as a consumable library? If the actual code for performing the delete were wrapped up in a module, buildpacks like Java could consume the module and integrate it precisely where needed.

Reply:
I suppose another possibility would be to have this RFC document the interface and env variables that should be exposed & to contribute the buildpack as an implementation of this interface. That would free buildpacks like Java to implement it in their own way, but follow the prescribed interface so the user experience is the same.

Reply:
I think the real issue you have highlighted is that having some way for the buildpacks to communicate to the source removal buildpack would be very beneficial. 

I think where I run into problems with that idea is that this buildpack also accepts user input as to what should stay and what should go and it seems like an extremely difficult problem trying to reconcile the user inputs away for the buildpack inputs. This buildpack is more of a way for user to configure what of their own files do they want on the final image. It seems counter intuitive to me at least to make this a buildpack leveraged by other buildpacks.

However, I think you did purpose an interesting having cake and eating it too scenario where we rewrite the buildpack slightly to be a wrapper for a removal library. That way buildpacks can benefit from the logic but we are not blurring the lines as to the purpose of the source removal buildpack.

This means that there is still a buildpack that gives users tight control over the files in the workspace that end up in the final image and buildpack authors get the benefit of shared complex logic.

I would love to hear other peoples thoughts on this!

Reply:
Again I think that these are changes that could be made after adoption as the overall user facing API for the buildpack would not change if we refactored the buildpack to be a wrapper for a library.

Reply:
I think that is a fair addition and sorry for confusion!

Reply:
Can we discuss a bit more the concerns about having a buildpack set or modify the values of `BP_INCLUDE_FILES` and `BP_EXCLUDE_FILES`? It feels reasonable to me that a buildpack would do this so that it could signal downstream to the `source-removal` buildpack what should be kept or removed.

For example, I don't see why you couldn't solve the proposed scenario outlined by @dmikusa-pivotal using this API:

> At a high level:
>
>    1. Do a build, the artifact is stored in a layer so the build is cached
>    2. We delete source code
>    3. Artifact is taken from layer and moved to /workspace

I don't see why that couldn't be the following:

1. Do a build, the artifact is stored in a layer so the build is cached
2. Artifact is taken from layer and moved to /workspace
3. `BP_INCLUDE_FILES` is modified to include the artifact
4. `source-removal` buildpack clears the /workspace of everything except the artifact

I'm probably missing something. What's the concern here?

Reply:
I guess that is fair. The buildpack could append included files. I think that it gets tricky if a uses wants to use exclude instead there would be so required logic to do both which wouldn't be impossible it is just not implemented.

Reply:
I have modified the buildpack so that it can run while both environment variables are set. It will first run the include files removal logic and then follow up with the exclude files removal logic. I think that this would give a fair pathway for buildpacks to ask for certain files to be included without interfering with the users environment variables